**NumPy -> Numerical Python**

Numpy is a Python Library where we have lot of Built_in_functions which are used to develop AI

In [1]:
import numpy as np

In [ ]:
len(dir(np))

532

**Out of 532 Built_in_functions in NumPy we need to learn the functions which are used in AI**

- `array `
- `ndim `
- `shape `
- `size `
- `arange `
- `reshape `
- `reshape (-1,1) and reshape (1,-1)`
- `zeros `
- `Ones `
- `full `
- `ravel`
- `concatination `
- `transpose `
- `diagonal `
- `random -> randint `
- `random -> randn`
- `random -> rand`
- `random -> random `
- `max `
- `min `
- `sum `
- `argmax `
- `argmin `
- `axis = 1`
- `axis = 0 `
- `maths functions `
- `matmul `
- `repeat `
- `vstack `
- `hstack `
- `linespace `


In [ ]:
# NumPy Library is used to create the data in the form of Dimensions
# or it is used to work with the data which is aready in the form of Dimensions

In [ ]:
a = [10,20,30,40]
print(a)

[10, 20, 30, 40]


In [ ]:
# array -> will create the data in the form of dimesion (1D | 2D | 3D)

a = np.array( [10 , 20 , 30,40] )

print(a) # calling from the Memory

print(a.ndim) # ndim (number of dimesions)

print(a.shape)  # shape (rows and columns)

[10 20 30 40]
1
(4,)


In [ ]:
# array -> will create the data in the form of dimesion (1D | 2D | 3D)

a = np.array( [10 , 20 , 30,40] )

print(a) # calling from the Memory

print(a.ndim) # ndim (number of dimesions)

print(a.shape)  # shape (rows and columns)

[10 20 30 40]
1
(4,)


In [ ]:
print(a[1])

20


In [ ]:
a = np.array([ [11,4,7] , [3,2,9] , [6,14,8] ])

print(a)

print(a.ndim)

print(a.shape) # (rows ,columns)

print(a.size) # total values

[[11  4  7]
 [ 3  2  9]
 [ 6 14  8]]
2
(3, 3)
9


In [ ]:
print(a[2][2])

8


In [ ]:
print(a[1][2])

9


In [ ]:
# arange will work like range in forloop only

a = np.arange(11,17) # it will work with (n-1) and the output will be in the form on 1D
print(a)
print(a.ndim)

[11 12 13 14 15 16]
1


In [ ]:
print(a[3])

14


In [ ]:
# 1D to 2D

a = np.arange(11,17)
print(a)
print(a.ndim)
print(a.shape)

print('=========================')

b = a.reshape(2,3)
print(b)
print(b.ndim)
print(b.shape)

[11 12 13 14 15 16]
1
(6,)
[[11 12 13]
 [14 15 16]]
2
(2, 3)


In [ ]:
a = np.arange(11,17).reshape(2,3)
print(a)
print(a.ndim)
print(a.shape)

[[11 12 13]
 [14 15 16]]
2
(2, 3)


In [ ]:
a = np.arange(1,61).reshape(10,6)
print(a)

[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]
 [25 26 27 28 29 30]
 [31 32 33 34 35 36]
 [37 38 39 40 41 42]
 [43 44 45 46 47 48]
 [49 50 51 52 53 54]
 [55 56 57 58 59 60]]


In [ ]:
print(a[3][2])

21


In [ ]:
# 1D to 3D

a = np.arange(1,61).reshape(5,2,6)
print(a)

[[[ 1  2  3  4  5  6]
  [ 7  8  9 10 11 12]]

 [[13 14 15 16 17 18]
  [19 20 21 22 23 24]]

 [[25 26 27 28 29 30]
  [31 32 33 34 35 36]]

 [[37 38 39 40 41 42]
  [43 44 45 46 47 48]]

 [[49 50 51 52 53 54]
  [55 56 57 58 59 60]]]


In [ ]:
print(a[3] [0] [5])

42


In [ ]:
print(a[1] [1] [3])

22


In [ ]:
# 1D to 3D

a = np.arange(1,61).reshape(2,5,6)
print(a)

[[[ 1  2  3  4  5  6]
  [ 7  8  9 10 11 12]
  [13 14 15 16 17 18]
  [19 20 21 22 23 24]
  [25 26 27 28 29 30]]

 [[31 32 33 34 35 36]
  [37 38 39 40 41 42]
  [43 44 45 46 47 48]
  [49 50 51 52 53 54]
  [55 56 57 58 59 60]]]


In [ ]:
a.ndim

3

In [ ]:
print(a[0] [2] [3])

16


In [ ]:
print(a[1] [3]  [4])

53


In [ ]:
# reshape(-1,1)

a = np.array([10,11,12,13])
print(a)
print(a.ndim)
print(a.shape)

print("==========================")

b = a.reshape(-1,1)
print(b)
print(b.ndim)
print(b.shape)

[10 11 12 13]
1
(4,)
[[10]
 [11]
 [12]
 [13]]
2
(4, 1)


In [ ]:
# reshape(1,-1)

a = np.array([10,11,12,13])
print(a)
print(a.ndim)
print(a.shape)

print("==========================")

b = a.reshape(1,-1)
print(b)
print(b.ndim)
print(b.shape)

[10 11 12 13]
1
(4,)
[[10 11 12 13]]
2
(1, 4)


In [4]:
# zeros -> create zeros for specified rows and columns
a = np.zeros( (5,6)   )

a

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

In [5]:
a.dtype

dtype('float64')

In [6]:
# zeros -> create zeros for specified rows and columns
a = np.zeros( (5,6)  ,dtype = np.int64 )

a

array([[0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0]])

In [7]:
# ones
a = np.ones( (5,6) )

a

array([[1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1.]])

In [9]:
# full

a = np.full( (5 , 10) , 7)
a

array([[7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
       [7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
       [7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
       [7, 7, 7, 7, 7, 7, 7, 7, 7, 7],
       [7, 7, 7, 7, 7, 7, 7, 7, 7, 7]])

In [12]:
# numpy -> random -> randint
# give range it will pick random integer values

a = np.random.randint(10,50 , 5)

a

array([34, 43, 28, 19, 16])

In [14]:
# numpy -> random -> rand
# rand -> function will gives values between 0 to 1

a = np.random.rand(4,4)
a

array([[0.52257886, 0.35829013, 0.50699057, 0.0983507 ],
       [0.71013418, 0.16553625, 0.30896016, 0.70456857],
       [0.28214492, 0.81382324, 0.97853482, 0.13130175],
       [0.31604748, 0.63066992, 0.85993823, 0.88514374]])

In [15]:
# numpy -> random -> rand
# random -> function will gives values between 0 to 1

a = np.random.random((4,4))
a

array([[0.47273786, 0.19202914, 0.72595637, 0.67071679],
       [0.8977339 , 0.48413809, 0.96276512, 0.369672  ],
       [0.13193956, 0.10509527, 0.31540815, 0.05245209],
       [0.75547277, 0.23280614, 0.94341454, 0.15340371]])

In [16]:
# numpy -> random -> randn
# randn -> function will gives values between -3 to 3

a = np.random.randn(4,4)
a

array([[ 1.069257  , -0.76364534,  0.78739989, -0.60783531],
       [-2.18633329,  0.39683755, -0.43070269, -1.32978111],
       [-0.31471021,  0.38395876, -0.27227603, -0.68326901],
       [ 1.95544516, -0.02135351, -0.38716608,  0.41695322]])

In [18]:
# ravel -> any dimension data into 1D

a = np.array([  [10,9,8] , [7,4,11] , [13,2,6]   ])
print(a)
print(a.ndim)
print(a.shape)
print(a.size)

print('===================')

b = np.ravel(a)
print(b)
print(b.ndim)
print(b.shape)

[[10  9  8]
 [ 7  4 11]
 [13  2  6]]
2
(3, 3)
9
[10  9  8  7  4 11 13  2  6]
1
(9,)


In [19]:
# transpose
a = np.array([  [10,9,8] , [7,4,11] , [13,2,6]   ])
print(a)

print("===================")

b = np.transpose(a)
print(b)

[[10  9  8]
 [ 7  4 11]
 [13  2  6]]
[[10  7 13]
 [ 9  4  2]
 [ 8 11  6]]


In [22]:
# concatinate
a = np.array([  [10,9,8] , [7,4,11] , [13,2,6]   ])
print(a)
print('=============================')
b = np.array([ [1,2,3] , [4,5,6] , [10,9,8]  ])
print(b)

print('================================')

c = np.concatenate((a,b))
print(c)

[[10  9  8]
 [ 7  4 11]
 [13  2  6]]
[[ 1  2  3]
 [ 4  5  6]
 [10  9  8]]
[[10  9  8]
 [ 7  4 11]
 [13  2  6]
 [ 1  2  3]
 [ 4  5  6]
 [10  9  8]]


In [23]:
# axis = 1 -> means -> row
# axis = 0 -> means -> column

In [24]:
a = np.array([  [10,9,8] , [7,4,11] , [13,2,6]   ])
print(a)

[[10  9  8]
 [ 7  4 11]
 [13  2  6]]


In [26]:
np.sum(a)

np.int64(70)

In [27]:
np.sum(a , axis=1)

array([27, 22, 21])

In [28]:
np.sum(a , axis=0)

array([30, 15, 25])

In [29]:
np.max(a)

np.int64(13)

In [30]:
np.max(a , axis=1)

array([10, 11, 13])

In [31]:
np.max(a , axis=0)

array([13,  9, 11])

In [34]:
np.min(a , axis=0)

array([7, 2, 6])

In [35]:
np.max(a)

np.int64(13)

In [36]:
np.argmax(a)

np.int64(6)

In [37]:
np.max(a , axis=1)


array([10, 11, 13])

In [38]:
np.argmax(a , axis=1)

array([0, 2, 0])

In [39]:
np.min(a)

np.int64(2)

In [40]:
np.argmin(a)

np.int64(7)

In [41]:
np.min(a , axis=1)

array([8, 4, 2])

In [42]:
np.argmin(a , axis=1)

array([2, 1, 1])

In [43]:
np.argmin(a , axis=0)

array([1, 2, 2])

In [44]:
# math operations
a = np.array([ [1,2] , [3,4] ])
b = np.array([ [10,20] ,[30,40] ])

np.add(a,b)

array([[11, 22],
       [33, 44]])

In [45]:
np.subtract(a,b)

array([[ -9, -18],
       [-27, -36]])

In [46]:
# matmul

a = np.array( [  [1,2] , [3,4] ])
b = np.array( [ [10,11,12] , [3,8,7] ])

np.matmul(a,b)

array([[16, 27, 26],
       [42, 65, 64]])

In [47]:
# matmul

a = np.array( [  [1,2] , [3,4] ])
b = np.array( [ [10,11,12] , [3,8,7] ,[21,22,23]])

np.matmul(a,b)

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 2)

In [51]:
# linespace

a = np.linspace(10,25 , 6)
a

array([10., 13., 16., 19., 22., 25.])

In [54]:
# diagonal
a = np.array([  [10,9,8] , [7,4,11] , [13,2,6]   ])

np.diagonal(a)

array([10,  4,  6])

In [55]:
a = np.array([[1,2] , [3,4]])
b = np.array([[10,11],[12,13]])

np.vstack((a,b))

array([[ 1,  2],
       [ 3,  4],
       [10, 11],
       [12, 13]])

In [56]:
a = np.array([[1,2] , [3,4]])
b = np.array([[10,11],[12,13]])

np.hstack((a,b))

array([[ 1,  2, 10, 11],
       [ 3,  4, 12, 13]])

In [61]:
# repeat

a = np.array([   [10,12,14,16]   ])

a

array([[10, 12, 14, 16]])

In [63]:
np.repeat(a , 5 , axis = 0)

array([[10, 12, 14, 16],
       [10, 12, 14, 16],
       [10, 12, 14, 16],
       [10, 12, 14, 16],
       [10, 12, 14, 16]])